pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+func GetAuthentication(registry string) (string, string, error) {
+	registry = DockerHostNameToRegistry(registry)
+	// Check auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	username, password, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if username != "" && password != "" {
+		return username, password, nil
+	}
+
+	// check ~/.docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	username, password, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, "", "", path, deleteAuth)
+}
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0700); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file and returns it
+// or returns an empty dockerConfigFile data structure if auth.json is empty
+func readJSONFile(path string) (dockerConfigFile, error) {
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if err = json.Unmarshal(raw, &auths); err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+	}
+
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func modifyJSON(registry, username, password, path, action string) error {
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	switch action {
+	case addAuth:
+		if len(auths.CredHelpers) > 0 {
```

Comment:
Shouldn’t this also be
```go
if ch, exists := auths.CredHelpers[registry]; exists
```
?

Reply:
I thought the user just had to manually enter one entry, and then we can keep adding to it. It is your call if you want the user to enter the keys, and config.go will just add the authentication to it.

Reply:
> I thought the user just had to manually enter one entry, and then we can keep adding to it.

It’s plausible that if the user wants to use a helper for one registry, it may be useful to use the same helper for other registries; but what if e.g. `CredHelpers` is already configured for 3 registry hostnames, to use two different helpers? I can’t see how code can decide which of the two helpers to use when the user refers to a 4th registry.

So, it seems to me that forcing users to manually configure a hostname→helper mapping for every hostname they care about is the only practical solution (other than an UI callback to ask the user, I guess). That will work well if the helper is somehow specialized for that hostname (e.g. a cloud-provider-specific helper to obtain an expiring token). If the user wanted to use a single helper for everything (like gnome-keyring), it does not work well at all—we would probably want to implement `credsStore`; but that’s really a quite separate feature.

Or maybe I am missing something about the design of the pre-existing implementation.

Reply:
hmm, I agree. I think implementing credStore in future would be great as well.

Reply:
fixed